src/drivers/wayland/wayland.c

```diff
@@ -0,0 +1,2565 @@
+        goto err_free;
+    }
+
+    obj->buffer_group = smm_create();
+    if (obj->buffer_group == NULL) {
+        LV_LOG_ERROR("cannot create buffer group for graphic object");
+        goto err_destroy_surface;
+    }
+
+    obj->window = window;
+    obj->type = type;
+    obj->surface_configured = true;
+    obj->pending_buffer = NULL;
+    wl_surface_set_user_data(obj->surface, obj);
+    SMM_TAG(obj->buffer_group, LW_TAG_LOCAL, obj);
+
+    return obj;
+
+err_destroy_surface:
+    wl_surface_destroy(obj->surface);
+
+err_free:
+    lv_free(obj);
+err_out:
+
+    return NULL;
+}
+
+static void destroy_wl_surface_obj(struct lw_surface_object* obj)
+{
+    if (obj->subsurface) {
+        wl_subsurface_destroy(obj->subsurface);
+    }
+
+    wl_surface_destroy(obj->surface);
+    smm_destroy(obj->buffer_group);
+    lv_free(obj);
+}
+
+static bool resize_window(lw_window_t* window, int width, int height)
+{
+    lv_color32_t* buf1 = NULL;
+    struct smm_buffer_t* body_buf1;
+    struct smm_buffer_t* body_buf2;
+    int b;
+
+    LV_LOG_TRACE("resize window %dx%d frame: %ld rendered: %d",
+        width, height, window->frame_counter, window->frame_done);
+
+/* Update size for newly allocated buffers */
+    smm_resize(window->body->buffer_group, ((width * LW_COLOR_PX_SIZE) * height) * 2);
+
+    window->width = width;
+    window->height = height;
+
+    window->body->width = width;
+    window->body->height = height;
+
+    /* Pre-allocate two buffers for the window body here */
+    body_buf1 = smm_acquire(window->body->buffer_group);
+    body_buf2 = smm_acquire(window->body->buffer_group);
+
+    if (smm_map(body_buf2) == NULL) {
+        LV_LOG_ERROR("Cannot pre-allocate backing buffers for window body");
+        wl_surface_destroy(window->body->surface);
+        return false;
+    }
+
+    /* Moves the buffers to the the unused list of the group */
+    smm_release(body_buf1);
+    smm_release(body_buf2);
+
+#if LV_WAYLAND_CLIENT_SIDE_DECORATIONS
+    if (!window->application->opt_disable_decorations && !window->fullscreen) {
+        for (b = 0; b < LW_NUM_DECORATIONS; b++) {
+            if (!create_wl_decoration(window, window->decoration[b],
+                window->body->width, window->body->height, 0x00)) {
+                LV_LOG_ERROR("failed to create decoration %d", b);
+            }
+        }
+    }
+    else if (!window->application->opt_disable_decorations) {
+        /* Entering fullscreen, detach decorations to prevent xdg_wm_base error 4 */
+        /* requested geometry larger than the configured fullscreen state */
+        for (b = 0; b < LW_NUM_DECORATIONS; b++) {
+            detach_wl_decoration(window, window->decoration[b]);
+        }
+    }
+#endif
+
+    if (window->display != NULL) {
+        /* Resize draw buffer */
+        uint32_t data_size = ((width * height) / LW_DRAW_BUFFER_DIV) * sizeof(lv_color32_t);
+        buf1 = lv_malloc_zeroed(data_size);
```

Comment:
Use normal `malloc` here to avoid issues with too small LV_MEM_SIZE.

Reply:
Should it be? I thought to take advantage of it.